File: ios/Sources/ILDSDesignSystem/IldsSearch.swift
import ILDSTokens
import SwiftUI

// Mirrors lib/ilds_search.dart + web Search.tsx (Figma 13965:16190).

public struct IldsSearch: View {
    @Binding private var text: String
    private let placeholder: String
    private let onSubmit: ((String) -> Void)?
    private let onClear: (() -> Void)?
    private let isLoading: Bool
    private let isDisabled: Bool

    @FocusState private var isFocused: Bool

    public init(
        text: Binding<String>,
        placeholder: String = "Search",
        onSubmit: ((String) -> Void)? = nil,
        onClear: (() -> Void)? = nil,
        isLoading: Bool = false,
        isDisabled: Bool = false
    ) {
        _text = text
        self.placeholder = placeholder
        self.onSubmit = onSubmit
        self.onClear = onClear
        self.isLoading = isLoading
        self.isDisabled = isDisabled
    }

    public var body: some View {
        let colors = IldsSearchColors.resolve(isFocused: isFocused, isDisabled: isDisabled)

        HStack(spacing: ILDSTokens.sp8) {
            Image(systemName: "magnifyingglass")
                .font(.system(size: ILDSTokens.sp16))
                .foregroundStyle(colors.leadingIcon)
            TextField(placeholder, text: $text)
                .font(.system(size: ILDSTokens.fontSize14, weight: ILDSTokens.fontWeightRegular))
                .foregroundStyle(isDisabled ? ILDSTokens.neutralCoolgray300 : ILDSTokens.neutralCoolgray900)
                .focused($isFocused)
                .disabled(isDisabled || isLoading)
                .onSubmit { onSubmit?(text) }
            if isLoading {
                ProgressView()
                    .progressViewStyle(.circular)
                    .tint(ILDSTokens.primaryOrange500)
                    .frame(width: ILDSTokens.sp20, height: ILDSTokens.sp20)
            } else if !text.isEmpty, !isDisabled {
                Button(action: clear) {
                    Image(systemName: "xmark")
                        .font(.system(size: ILDSTokens.sp16))
                        .foregroundStyle(ILDSTokens.neutralCoolgray500)
                }
                .buttonStyle(.plain)
                .accessibilityLabel("Clear search")
            }
        }
        .padding(.horizontal, ILDSTokens.sp12)
        .frame(height: 44)
        .background(colors.background)
        .overlay {
            RoundedRectangle(cornerRadius: ILDSTokens.radiusMedium)
                .strokeBorder(colors.border, lineWidth: colors.borderWidth)
        }
        .clipShape(RoundedRectangle(cornerRadius: ILDSTokens.radiusMedium))
        .accessibilityLabel(placeholder)
    }

    private func clear() {
        text = ""
        onClear?()
    }
}

private struct IldsSearchColors {
    let border: Color
    let background: Color
    let leadingIcon: Color
    let borderWidth: CGFloat

    static func resolve(isFocused: Bool, isDisabled: Bool) -> IldsSearchColors {
        if isDisabled {
            return IldsSearchColors(
                border: ILDSTokens.neutralCoolgray300,
                background: ILDSTokens.neutralCoolgray200,
                leadingIcon: ILDSTokens.neutralCoolgray300,
                borderWidth: 1
            )
        }
        let border = isFocused ? ILDSTokens.primaryOrange600 : ILDSTokens.neutralCoolgray500
        let borderWidth: CGFloat = isFocused ? 2 : 1
        let leadingIcon = isFocused ? ILDSTokens.primaryOrange500 : ILDSTokens.neutralCoolgray400
        return IldsSearchColors(
            border: border,
            background: ILDSTokens.neutralCoolgray50,
            leadingIcon: leadingIcon,
            borderWidth: borderWidth
        )
    }
}

File: android/ilds-design-system/src/main/kotlin/com/icicilombard/ilds/components/IldsSearch.kt
package com.icicilombard.ilds.components

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.interaction.collectIsFocusedAsState
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.Search
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.icicilombard.ilds.tokens.IldsTokens

// Figma set 13965:16190 — mirrors web/Search.tsx + lib/ilds_search.dart.

@Composable
fun IldsSearch(
    modifier: Modifier = Modifier,
    value: String = "",
    onValueChange: (String) -> Unit = {},
    placeholder: String = "Search",
    isLoading: Boolean = false,
    isDisabled: Boolean = false,
    semanticLabel: String = "Search",
    onSubmit: (String) -> Unit = {},
    onClear: () -> Unit = {},
) {
    var text by remember(value) { mutableStateOf(value) }
    val interactionSource = remember { androidx.compose.foundation.interaction.MutableInteractionSource() }
    val isFocused by interactionSource.collectIsFocusedAsState()
    val hasText = text.isNotEmpty()

    val borderColor = when {
        isDisabled -> IldsTokens.neutralCoolgray300
        isFocused -> IldsTokens.primaryOrange600
        hasText -> IldsTokens.neutralCoolgray500
        else -> IldsTokens.neutralCoolgray500
    }
    val borderWidth = if (isFocused && !isDisabled) 2.dp else 1.dp
    val background = when {
        isDisabled -> IldsTokens.neutralCoolgray200
        isFocused -> IldsTokens.neutralCoolgray50
        else -> IldsTokens.neutralCoolgray50
    }
    val iconColor = when {
        isDisabled -> IldsTokens.neutralCoolgray400
        isFocused -> IldsTokens.primaryOrange500
        else -> IldsTokens.neutralCoolgray400
    }

    Surface(
        modifier = modifier
            .fillMaxWidth()
            .semantics { contentDescription = semanticLabel },
        shape = RoundedCornerShape(IldsTokens.radiusMedium),
        color = background,
        border = BorderStroke(borderWidth, borderColor),
    ) {
        Row(
            modifier = Modifier
                .defaultMinSize(minHeight = 44.dp)
                .padding(horizontal = IldsTokens.sp12),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Icon(
                imageVector = Icons.Filled.Search,
                contentDescription = null,
                tint = iconColor,
                modifier = Modifier.size(16.dp),
            )
            BasicTextField(
                value = text,
                onValueChange = {
                    text = it
                    onValueChange(it)
                },
                modifier = Modifier
                    .weight(1f)
                    .padding(horizontal = IldsTokens.sp8),
                enabled = !isDisabled && !isLoading,
                singleLine = true,
                textStyle = TextStyle(
                    fontSize = IldsTokens.fontSize14,
                    lineHeight = 18.sp,
                    fontWeight = IldsTokens.fontWeightRegular,
                    color = IldsTokens.neutralCoolgray900,
                ),
                cursorBrush = SolidColor(IldsTokens.primaryOrange500),
                interactionSource = interactionSource,
                keyboardOptions = KeyboardOptions(imeAction = ImeAction.Search),
                keyboardActions = KeyboardActions(onSearch = { onSubmit(text) }),
                decorationBox = { inner ->
                    Box(contentAlignment = Alignment.CenterStart) {
                        if (text.isEmpty()) {
                            Text(
                                text = placeholder,
                                style = TextStyle(
                                    fontSize = IldsTokens.fontSize14,
                                    lineHeight = 18.sp,
                                    fontWeight = IldsTokens.fontWeightRegular,
                                ),
                                color = IldsTokens.neutralCoolgray500,
                            )
                        }
                        inner()
                    }
                },
            )
            when {
                isLoading -> CircularProgressIndicator(
                    modifier = Modifier.size(20.dp),
                    color = IldsTokens.primaryOrange500,
                    strokeWidth = 2.dp,
                )
                hasText && !isDisabled -> IconButton(
                    onClick = {
                        text = ""
                        onValueChange("")
                        onClear()
                    },
                    modifier = Modifier.size(24.dp),
                ) {
                    Icon(
                        imageVector = Icons.Filled.Close,
                        contentDescription = "Clear search",
                        tint = IldsTokens.neutralCoolgray500,
                        modifier = Modifier.size(16.dp),
                    )
                }
            }
        }
    }
}
